File: cmd/app/service_windows.go
//go:build windows

package main

import (
	"fmt"
	"strings"

	"golang.org/x/sys/windows/registry"
)

// runKeyName is the registry value name under HKCU\...\Run used to autostart
// the daemon at login.
const runKeyName = "LocalAgent"

// runKeyPath is the per-user Run key under HKEY_CURRENT_USER. A value here
// starts the referenced program at login without requiring admin privileges.
const runKeyPath = `Software\Microsoft\Windows\CurrentVersion\Run`

// runKeyValue builds the registry value that launches `<binary> start` at
// login. The binary path is quoted so paths containing spaces are handled
// correctly by the Windows shell. Extracted for testability.
func runKeyValue(binary string) string {
	// Quote the executable path so a path with spaces (e.g.
	// "C:\Program Files\app.exe") is parsed as a single argument.
	return fmt.Sprintf(`"%s" start`, binary)
}

// installService registers the daemon to start at login via the per-user Run
// registry key. This is the simplest reliable user-level autostart on Windows
// and avoids the admin privileges a true Windows Service would require.
func installService(user bool) error {
	if !user {
		return fmt.Errorf("system-wide install is not supported on Windows; use --user (the default)")
	}

	binary, err := resolveBinaryPath()
	if err != nil {
		return err
	}

	// Open the user Run key for writing. HKCU does not require admin rights.
	key, err := registry.OpenKey(registry.CURRENT_USER, runKeyPath, registry.SET_VALUE|registry.QUERY_VALUE)
	if err != nil {
		return fmt.Errorf("open HKCU\\%s: %w", runKeyPath, err)
	}
	defer func() { _ = key.Close() }()

	// Refuse to silently overwrite an existing entry.
	if existing, _, err := key.GetStringValue(runKeyName); err == nil && existing != "" {
		return fmt.Errorf("autostart entry already exists (HKCU\\%s\\%s = %q) — run 'app uninstall-service' first",
			runKeyPath, runKeyName, existing)
	} else if err != nil && err != registry.ErrNotExist {
		// Some other read error; surface it rather than guessing.
		return fmt.Errorf("read existing run key: %w", err)
	}

	value := runKeyValue(binary)
	if err := key.SetStringValue(runKeyName, value); err != nil {
		return fmt.Errorf("set run key value: %w", err)
	}

	fmt.Printf("Installed autostart entry: HKCU\\%s\\%s = %s\n", runKeyPath, runKeyName, value)
	return nil
}

// uninstallService removes the per-user Run registry entry created by
// installService.
func uninstallService(user bool) error {
	if !user {
		return fmt.Errorf("system-wide uninstall is not supported on Windows; use --user (the default)")
	}

	key, err := registry.OpenKey(registry.CURRENT_USER, runKeyPath, registry.SET_VALUE|registry.QUERY_VALUE)
	if err != nil {
		return fmt.Errorf("open HKCU\\%s: %w", runKeyPath, err)
	}
	defer func() { _ = key.Close() }()

	if err := key.DeleteValue(runKeyName); err != nil {
		if strings.Contains(err.Error(), "RegDeleteValue") {
			// The value did not exist; treat as success so uninstall is idempotent.
			fmt.Println("Autostart entry was not present; nothing to remove.")
			return nil
		}
		return fmt.Errorf("delete run key value: %w", err)
	}

	fmt.Printf("Removed autostart entry: HKCU\\%s\\%s\n", runKeyPath, runKeyName)
	return nil
}

File: src/cli/service_windows.rs
//! Windows per-user Run-key integration.
//!
//! A true Windows Service needs elevated installation and a service manager
//! lifecycle. The current-user Run key is the intentional, non-admin
//! equivalent of the Linux/macOS per-user service implementations.

use std::io::{self, Write};

use anyhow::{anyhow, Context, Result};
use tracing::info;
use winreg::enums::{HKEY_CURRENT_USER, KEY_QUERY_VALUE, KEY_SET_VALUE};
use winreg::RegKey;

const RUN_KEY_NAME: &str = "LocalAgent";
const RUN_KEY_PATH: &str = r"Software\Microsoft\Windows\CurrentVersion\Run";

/// Quote the executable and append the foreground daemon subcommand.
fn run_key_value(binary: &str) -> String {
    format!(r#""{binary}" start"#)
}

/// Open the user's Run key with the minimal access required for this command.
fn run_key() -> Result<RegKey> {
    RegKey::predef(HKEY_CURRENT_USER)
        .open_subkey_with_flags(RUN_KEY_PATH, KEY_QUERY_VALUE | KEY_SET_VALUE)
        .with_context(|| format!(r"open HKCU\{RUN_KEY_PATH}"))
}

/// Register the daemon to start after the current user logs in.
pub(super) fn install() -> Result<()> {
    let binary = std::env::current_exe().context("resolve current executable")?;
    let key = run_key()?;
    match key.get_value::<String, _>(RUN_KEY_NAME) {
        Ok(existing) if !existing.is_empty() => {
            return Err(anyhow!(
                r"autostart entry already exists (HKCU\{RUN_KEY_PATH}\{RUN_KEY_NAME} = {existing:?}) \
                 — run 'local_agent uninstall-service' first"
            ));
        }
        Ok(_) => {}
        Err(error) if error.kind() == io::ErrorKind::NotFound => {}
        Err(error) => return Err(error).context("read existing Run-key value"),
    }

    let value = run_key_value(&binary.to_string_lossy());
    key.set_value(RUN_KEY_NAME, &value)
        .context("set Run-key value")?;

    info!(key = %RUN_KEY_PATH, value, "installed Windows Run-key autostart");
    writeln!(
        io::stdout(),
        r"Installed autostart entry: HKCU\{RUN_KEY_PATH}\{RUN_KEY_NAME} = {value}"
    )
    .context("write service install status")
}

/// Remove the daemon's per-user autostart entry.
pub(super) fn uninstall() -> Result<()> {
    let key = run_key()?;
    match key.get_value::<String, _>(RUN_KEY_NAME) {
        Ok(_) => {}
        Err(error) if error.kind() == io::ErrorKind::NotFound => {
            writeln!(
                io::stdout(),
                "Autostart entry was not present; nothing to remove."
            )
            .context("write service uninstall status")?;
            return Ok(());
        }
        Err(error) => return Err(error).context("read existing Run-key value"),
    }
    key.delete_value(RUN_KEY_NAME)
        .context("delete Run-key value")?;

    info!(key = %RUN_KEY_PATH, "removed Windows Run-key autostart");
    writeln!(
        io::stdout(),
        r"Removed autostart entry: HKCU\{RUN_KEY_PATH}\{RUN_KEY_NAME}"
    )
    .context("write service uninstall status")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn run_key_value_quotes_executable_paths() {
        assert_eq!(run_key_value(r"C:\app.exe"), r#""C:\app.exe" start"#);
        assert_eq!(
            run_key_value(r"C:\Program Files\local_agent.exe"),
            r#""C:\Program Files\local_agent.exe" start"#
        );
    }
}
